File: frontend/src/api/clients.ts
import apiClient from "./client"
import type { Client, PaginatedResponse } from "./types"

export interface ClientFilters {
  page?: number
  per_page?: number
  search?: string
  status?: string
  owner_id?: string
  tag_id?: string
  sort_by?: string
  sort_order?: string
}

export interface ClientCreateData {
  name: string
  email?: string
  phone?: string
  company?: string
  industry?: string
  website?: string
  address?: string
  status?: string
  source?: string
  tag_ids?: string[]
}

export type ClientUpdateData = Partial<ClientCreateData>

export async function getClients(filters: ClientFilters = {}): Promise<PaginatedResponse<Client>> {
  const res = await apiClient.get<PaginatedResponse<Client>>("/clients", { params: filters })
  return res.data
}

export async function getClient(id: string): Promise<Client> {
  const res = await apiClient.get<Client>(`/clients/${id}`)
  return res.data
}

export async function createClient(data: ClientCreateData): Promise<Client> {
  const res = await apiClient.post<Client>("/clients", data)
  return res.data
}

export async function updateClient(id: string, data: ClientUpdateData): Promise<Client> {
  const res = await apiClient.put<Client>(`/clients/${id}`, data)
  return res.data
}

export async function deleteClient(id: string): Promise<void> {
  await apiClient.delete(`/clients/${id}`)
}

File: backend/app/api/v1/clients.py
"""Роутер клиентов — CRUD с owner-based доступом."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import get_current_user, require_role
from app.database import get_db
from app.models.user import User
from app.schemas.client import ClientCreate, ClientUpdate
from app.services import client_service

router = APIRouter(prefix="/clients", tags=["clients"])


@router.get("")
async def list_clients(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    search: str | None = None,
    status: str | None = None,
    owner_id: UUID | None = None,
    tag_id: UUID | None = None,
    sort_by: str = Query("created_at", pattern="^(name|created_at|updated_at)$"),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Список клиентов с фильтрацией и пагинацией."""
    return await client_service.list_clients(
        user, db, page, per_page, search, status, owner_id, tag_id, sort_by, sort_order
    )


@router.get("/{client_id}")
async def get_client(
    client_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Клиент по ID."""
    return await client_service.get_client(client_id, user, db)


@router.post("", status_code=201)
async def create_client(
    data: ClientCreate,
    user: User = Depends(require_role("admin", "manager")),
    db: AsyncSession = Depends(get_db),
):
    """Создание клиента."""
    return await client_service.create_client(data, user, db)


@router.put("/{client_id}")
async def update_client(
    client_id: UUID,
    data: ClientUpdate,
    user: User = Depends(require_role("admin", "manager")),
    db: AsyncSession = Depends(get_db),
):
    """Обновление клиента."""
    return await client_service.update_client(client_id, data, user, db)


@router.delete("/{client_id}")
async def delete_client(
    client_id: UUID,
    user: User = Depends(require_role("admin", "manager")),
    db: AsyncSession = Depends(get_db),
):
    """Удаление клиента."""
    return await client_service.delete_client(client_id, user, db)
